Feature localisation provider

matadam667 opened the pull request.

PaulinaCzapla committed:
Added map object to the scene.

PaulinaCzapla committed:
Implemented user location with automatic camera rotation by 90 degrees.

PaulinaCzapla committed:
Fixed camera rotation, it is now less sensitive to rapid changes of player rotation.

PaulinaCzapla committed:
Refactored code and changed camera rotation speed.

matadam667 committed:
Navigation

matadam667 committed:
Interface and implementation

matadam667 committed:
DeviceLocalizationProvider

Implementaiton of ILocalizationProvider

matadam667 committed:
FakeLocalizationProvider

matadam667 committed:
FakeLocalizationProvider

Works but some of the points are too close for the map to change localization - going to try to figue out how to work around it.

F4talErr0r commented on Assets/Mapbox/Unity/MeshGeneration/Modifiers/MeshModifiers/LineMeshModifier.cs:
```diff
@@ -23,7 +23,7 @@ public class LineMeshModifier : MeshModifier
 		#region Line Parameters
 
-		private LineGeometryOptions _options;
+		[SerializeField]private LineGeometryOptions _options;
```
This change was just for testing purposes, revert it.

F4talErr0r commented on Assets/Mapbox/Unity/MeshGeneration/Factories/DirectionsFactory.cs:
```diff
@@ -79,7 +79,7 @@ void Query()
 				wp[i] = _waypoints[i].GetGeoPosition(_map.CenterMercator, _map.WorldRelativeScale);
 			}
-			var _directionResource = new DirectionResource(wp, RoutingProfile.Driving);
+			var _directionResource = new DirectionResource(wp, RoutingProfile.Walking);
```
Try to write custom factory, don't modify original sources.

F4talErr0r commented on Assets/_Project/Scripts/Camera/CameraRotation.cs:
```diff
@@ -0,0 +1,78 @@
+using System;
```
Those are @PaulinaCzapla changes, so I won't discuss them here.

F4talErr0r commented on Assets/_Project/Scripts/Player/ILocalisationProvider.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace PolSl.UrbanHealthPath.Map
+{
+    public interface ILocalizationProvider
```
ILocationProvider would be more appropriate.

F4talErr0r replied:
Based on ILocationProvider from Mapbox, I think that we'll also need to receive both Coordinates and angle the user is heading, so this should probably return custom "Location" instance.

F4talErr0r commented on Assets/_Project/Scripts/Navigation/NavigationPointProvider.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace PolSl.UrbanHealthPath
+{
+    public class NavigationPointProvider : MonoBehaviour
```
What does it provide? I think it needs a complete rework. Maybe call it "PositionIndicatorRefresher", turn the method to UpdateIndicator(Vector3d position), create a class responsible for fetching the location data (TimedLocationFetcher) that would have LocationFetched event and subscribe PositionIndicatorRefresher to that event? Or something similiar.

F4talErr0r replied:
Nevermind, I just realized that it does completely different thing and is used for creating navigation path. TBH I don't think it's needed, we should consider a different approach (maybe NavigationManager?), but I'd leave it for now and refactor this functionality on a separate branch.

F4talErr0r commented on Assets/_Project/Scripts/Map/DeviceLocalizationProvider.cs:
```diff
@@ -0,0 +1,351 @@
+	/// This relies on Unity's <see href="https://docs.unity3d.com/ScriptReference/LocationService.html">LocationService</see> for location
+	/// and <see href="https://docs.unity3d.com/ScriptReference/Compass.html">Compass</see> for heading.
+	/// </summary>
+	public class DeviceLocalizationProvider : AbstractLocationProvider, ILocalizationProvider
```
Never just copy entire class from plugin to our own codebase. If we need to use it, create a wrapper that holds a reference.

F4talErr0r replied:
This class definitely does way too many things. Write a custom one and include only details that we really need.

F4talErr0r commented on Assets/_Project/Scripts/Map/AbstractFakeLocalizationProvider.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace PolSl.UrbanHealthPath.Map
+{
+	public abstract class AbstractFakeLocalizationProvider : AbstractLocationProvider
```
Do we really need this separation? I'd suggest leaving only ILocationProvider <- FakeLocationProvider, DeviceLocationProvider (unless those two really have something in common, then add custom AbstractLocationProvider with that common functionality).

matadam667 committed:
Factory + providers rewritten

Currently the map deosnt update. Working on it.

matadam667 committed:
Location Map Updater

matadam667 committed:
Factory + Updater + Smaller things refactored

Location finally works correctly - player is displayed exactly where we want them to be displayed.
